src/session.c

```diff
@@ -353,207 +516,157 @@ CK_RV p11prov_session_pool_init(P11PROV_CTX *ctx, CK_TOKEN_INFO *token,
         pool->sessions[i] = NULL;
     }
     OPENSSL_free(pool->sessions);
-    pool->sessions = NULL;
-    pool->num_p11sessions = 0;
-    pthread_mutex_unlock(&pool->lock);
-    /* ------------- LOCKED SECTION */
 
-    pthread_mutex_destroy(&pool->lock);
+    /* ------------- LOCKED SECTION */
+    (void)MUTEX_UNLOCK(pool);
+    (void)MUTEX_DESTROY(pool);
     OPENSSL_clear_free(pool, sizeof(P11PROV_SESSION_POOL));
-
-    return CKR_OK;
 }
 
 #define SESS_ALLOC_SIZE 32
-static P11PROV_SESSION *p11prov_session_new(P11PROV_CTX *ctx,
-                                            struct p11prov_slot *slot)
+
+/* NOTE: to be called with Pool Lock held */
+static P11PROV_SESSION *session_new(P11PROV_SESSION_POOL *pool)
 {
-    P11PROV_SESSION_POOL *p = slot->pool;
     P11PROV_SESSION *session;
+    int ret;
 
-    P11PROV_debug("Creating new P11PROV_SESSION session on pool %p", p);
+    P11PROV_debug("Creating new P11PROV_SESSION session on pool %p", pool);
 
     /* cap the amount of obtainable P11PROV_SESSIONs to double the max
      * number of available pkcs11 token sessions, just to have a limit
      * in case of runaway concurrent threads */
-    if (p->num_p11sessions > (int)(p->max_sessions * 2)) {
-        P11PROV_raise(ctx, CKR_SESSION_COUNT, "Max sessions limit reached");
+    if (pool->num_p11sessions > (int)(pool->max_sessions * 2)) {
+        P11PROV_raise(pool->provctx, CKR_SESSION_COUNT,
+                      "Max sessions limit reached");
         return NULL;
     }
 
     session = OPENSSL_zalloc(sizeof(P11PROV_SESSION));
     if (session == NULL) {
-        P11PROV_raise(ctx, CKR_HOST_MEMORY, "Failed to allocate session");
+        P11PROV_raise(pool->provctx, CKR_HOST_MEMORY,
+                      "Failed to allocate session");
         return NULL;
     }
-
-    session->provctx = ctx;
-    session->slotid = slot->id;
+    session->provctx = pool->provctx;
+    session->slotid = pool->slotid;
     session->session = CK_INVALID_HANDLE;
     session->flags = DEFLT_SESSION_FLAGS;
-    session->pool = p;
-    session->refcnt = 1;
+    session->pool = pool;
 
-    /* LOCKED SECTION ------------- */
-    pthread_mutex_lock(&p->lock);
+    ret = MUTEX_INIT(session);
+    if (ret != CKR_OK) {
+        OPENSSL_free(session);
+        return NULL;
+    }
 
     /* check if we need to expand the sessions array */
-    if ((p->num_p11sessions % SESS_ALLOC_SIZE) == 0) {
-        P11PROV_SESSION **tmp =
-            OPENSSL_realloc(p->sessions, (p->num_p11sessions + SESS_ALLOC_SIZE)
-                                             * sizeof(P11PROV_SESSION *));
+    if ((pool->num_p11sessions % SESS_ALLOC_SIZE) == 0) {
+        P11PROV_SESSION **tmp = OPENSSL_realloc(
+            pool->sessions, (pool->num_p11sessions + SESS_ALLOC_SIZE)
+                                * sizeof(P11PROV_SESSION *));
         if (tmp == NULL) {
-            P11PROV_raise(ctx, CKR_HOST_MEMORY,
+            P11PROV_raise(pool->provctx, CKR_CANT_LOCK,
```

Comment:
Should not this be the `CKR_HOST_MEMORY` error instead of the `CKR_CANT_LOCK`?

Reply:
Good catch!

Reply:
Pushed fix